### ISSUE TITLE
Tech discovery: strategies for minimizing BGS impact on Caseflow performance

### ISSUE THREAD

**Issue by laurjpeterson** [department-of-veterans-affairs/caseflow#6279]
# Improvements to BGS information in Queue

Caseflow is dependent upon BGS for information like Veteran name, POA, and appellant information. BGS can be unreliable, and Caseflow has the goal of remaining operational as much as possible even if external dependencies are experiencing downtime. Additionally, making multiple calls to BGS slows down Caseflow performance, as already experienced in our list of BEAAM appeals. 

Here, we try to think through improvements for how we might handle BGS data in Caseflow Queue.

## Overall Goals

- BVA employees should not be blocked from doing work during a BGS outage, wherever possible.
	- BVA employees should not need to wait for N BGS requests to resolve to view their list of N tasks.
	- BVA employees should be able to view case details for appeal info even if BGS is down.
- BVA employees should not be able to see veteran info from BGS if they lack permission to see that info.

## Queue task list view: Veteran name / appellant name

Currently, the Veteran name is currently displayed in all task list views, along with "Veteran is not the appellant" where applicable.

### Option 1. Don't show Veteran name/ "Veteran is not appellant"

This option would require redesigns that remove information from all Queues.

### Option 2. Lazy load Veteran names, don't show "Veteran is not appellant"

This would allow the Queue to load, but we would still need to make many BGS requests.

### Option 3. Save veteran name, a boolean representing whether the veteran is the appellant, and appellant name in the Caseflow database in Caseflow Intake 

If we get approval to save Veteran and appellant name in Caseflow, this is an option. Queue list view loading times will remain fast, and Veteran names change infrequently enough that data integrity wouldn't be an issue — according to Dustin, VACOLS veteran names are anecdotally not seen as a problem, even though they are not edited on any schedule.

## Case Details view: detailed veteran info

This includes gender, service periods, DOB, POA, address (what else)? See data in case details [here](https://github.com/department-of-veterans-affairs/caseflow/wiki/Case-details).

### Caching 

Some Veteran information is only accessible to users with certain sensitivity levels, so we'd likely need to cache this information on a per-user, per-appeal basis.

To ensure that the Board can continually review cases even through a BGS outage, we could run a nightly job to prefetch all relevant info from BGS for cases newly assigned to attorneys and judges. For other users, this information could just be fetched as necessary. 

The cached versions could be marked as expired after 24 hours, but if BGS is down we could still return the last saved version (while indicating that the information was updated more than a day ago). So only if a user has never looked at Case Details before for an appeal would they be possibly blocked by a BGS outage.

### Sensitivity

Option 1: Prevent users from loading the Case Details page if the user's BGS access level is too low.

Option 2: Display appeal information stored in Caseflow if the user's BGS access level is too low, but don't show BGS information.

### Open questions

- Should we lazy load BGS info in Case Details?
- Is the background job necessary, or should we fetch BGS info only when a user requests it?
- What other BGS data might we need - contested claimant info?

**Comment by mdbenjam** [comment]
Option 3 seems good for pulling data. Though, I think we should only save the veteran name in caseflow. From the earlier discussion we don't need to surface the appellant/who it is on the queue screen. As for sensitivity. I prefer option 1.

The background job may not be necessary if we cache when someone loads an appeal's case details page. It's possible just having the queue page available could be sufficient during an outage.

**Comment by laurjpeterson** [comment]
- **Queue task list view: Veteran name / appellant name**: I agree with option 3 with Mark's suggested edit. 
- **Sensitivity**: I prefer option 1 as well.
- **Cacheing**: That strategy looks good to me. I think having some Case details information is important during an outage. 
   - Separate but potentially related note/note for context, judges conducting hearings are very affected by Caseflow outages. If Caseflow is down, they will need to access their hearing worksheets in Hearing Prep, and have access to Veteran data so they can conduct the hearing.

**Comment by mkhandekar** [comment]
From the design side, then we'll need some sort of message / modal that appears when you click on a Veteran's name, telling them ... to view this case in VACOLS? 

Also, are we still thinking about improving the performance of the tables by adding pagination?

**Comment by amprokop** [comment]
> From the design side, then we'll need some sort of message / modal that appears when you click on a Veteran's name, telling them ... to view this case in VACOLS?

Hmm, i'm not sure I follow. When might we need this?

> Also, are we still thinking about improving the performance of the tables by adding pagination?

I believe not. The solution we've arrived at is to remove the need to make calls to BGS to load the tables. That should be enough for now.

**Comment by aroltsch** [comment]
Another option: we can only use cached version if BGS is down and throws an error. We can cache in Redis and expire after 24 hours.

**Comment by pkarman** [comment]
I think we've already done as much as we can to minimize BGS access.
